client/src/main/java/io/confluent/kafka/schemaregistry/client/CachedSchemaRegistryClient.java

```diff
@@ -474,25 +484,72 @@ public Collection<SubjectVersion> getAllVersionsById(int id) throws IOException,
   @Override
   public Schema getByVersion(String subject, int version, boolean lookupDeletedSchema) {
     try {
-      return restService.getVersion(subject, version, lookupDeletedSchema);
+      return getSchemaByVersion(subject, version, lookupDeletedSchema);
     } catch (IOException | RestClientException e) {
       throw new RuntimeException(e);
     }
   }
 
+  private Schema getSchemaByVersion(String subject, int version, boolean lookupDeletedSchema)
+      throws IOException, RestClientException {
+    final Map<Integer, Schema> versionSchemaMap = versionToSchemaCache.computeIfAbsent(
+        subject, k -> new BoundedConcurrentHashMap<>(cacheCapacity));
+
+    // The cache is only used when lookupDeletedSchema is true
+    Schema cachedSchema = lookupDeletedSchema ? versionSchemaMap.get(version) : null;
+    if (cachedSchema != null) {
+      return cachedSchema;
+    }
+
+    synchronized (this) {
+      cachedSchema = lookupDeletedSchema ? versionSchemaMap.get(version) : null;
```

Comment:
If we do this in synchronized mode, why do we still need to do it in line 499?

Reply:
This is the double-checked locking pattern. We first check outside of a synchronized block as it's more efficient.